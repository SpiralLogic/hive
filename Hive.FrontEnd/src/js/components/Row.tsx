import * as React from 'react';
import {Hexagon, HexCoordinates} from '../domain';
import CellMemo from './Cell';

const cellKey = ({q, r}: HexCoordinates) => `${q}-${r}`;
const cellComponent = (cell: Hexagon) => <CellMemo key={cellKey(cell.coordinates)} {...cell} />;
const emptyCell = (key: number) => <div key={key} className="hidden"/>;

type Props = {
    id: number;
    row: Array<Hexagon | false>;
};

function Row(props: Props) {
    const {row} = props;
    const cells = row.map((cell, i) => (cell && cellComponent(cell)) || emptyCell(i));
    return <div className="hex-row">{cells}</div>;
}

Row.displayName = 'Row';

function haveCellTilesBeenUpdated(prevProps: Props, newProps: Props) {
    return prevProps.row.every((cell, i) => newProps.row[i] && cell && (newProps.row[i] as Hexagon).tiles.length == (cell as Hexagon).tiles.length);
}

const RowMemo = React.memo(Row, (p, n) =>
    p.row.length == n.row.length &&
    haveCellTilesBeenUpdated(p, n));

export default RowMemo;
