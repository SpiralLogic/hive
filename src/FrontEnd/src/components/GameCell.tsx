import { FunctionComponent, h } from 'preact';
import { HexCoordinates, Tile as TileType } from '../domain';
import { MoveEvent, TileEvent } from '../utilities/hive-dispatcher';
import { addHiveDispatchListener, dispatchHiveEvent, useClassReducer } from '../utilities/hooks';
import { handleDragOver, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { useEffect, useState } from 'preact/hooks';
import Hexagon from './Hexagon';

type Props = { coords: HexCoordinates; hidden?: boolean };
const GameCell: FunctionComponent<Props> = (props) => {
  const { coords, children, hidden } = props;
  const [classes, setClasses] = useClassReducer('hide');
  const [selectedTile, setSelectedTile] = useState<TileType | null>(null);

  const isValidMove = (validMoves: HexCoordinates[]) =>
    validMoves.some((dest) => dest.q == coords.q && dest.r == coords.r);
  useEffect(() => setClasses({ type: hidden ? 'add' : 'remove', classes: ['hide'] }), [hidden]);

  addHiveDispatchListener<TileEvent>('tileDeselected', (event) => {
    if (!isValidMove(event.tile.moves)) {
      setClasses({ type: 'remove', classes: ['no-drop'] });
    }
    setSelectedTile(null);
    setClasses({ type: 'remove', classes: ['active', 'can-drop'] });
  });

  addHiveDispatchListener<TileEvent>('tileSelected', (event) => {
    if (!isValidMove(event.tile.moves)) {
      setClasses({ type: 'add', classes: ['no-drop'] });
    } else {
      setSelectedTile(event.tile);
      setClasses({ type: 'add', classes: ['can-drop'] });
    }
  });

  addHiveDispatchListener<TileEvent>('tileDropped', () => {
    if (classes.includes('active') && selectedTile && isValidMove(selectedTile.moves)) {
      dispatchHiveEvent<MoveEvent>({
        type: 'move',
        move: { coords, tileId: selectedTile.id },
      });
    }
  });

  const handleDragLeave = (event: DragEvent) => {
    event.stopPropagation();
    setClasses({ type: 'remove', classes: ['active'] });
  };

  const handleDragEnter = () => {
    setClasses({ type: 'add', classes: ['active'] });
  };

  const handleClick = (event: UIEvent) => {
    if (!(selectedTile && isValidMove(selectedTile.moves))) return;
    event.stopPropagation();
    dispatchHiveEvent({ type: 'move', move: { coords, tileId: selectedTile.id } });
    dispatchHiveEvent({ type: 'tileClear', tile: selectedTile });
  };

  const handleKeydown = (e: KeyboardEvent) => {
    if (!handleKeyboardNav(e) && isEnterOrSpace(e)) return handleClick(e);
  };

  const attributes = {
    class: classes || undefined,
    tabindex: selectedTile && isValidMove(selectedTile.moves) ? 0 : undefined,
    role: hidden ? 'none' : 'cell',
  };

  const handlers = {
    ondragover: handleDragOver,
    ondragleave: handleDragLeave,
    ondragenter: handleDragEnter,
    onmouseenter: handleDragEnter,
    onmouseleave: handleDragLeave,
    onclick: handleClick,
    onkeydown: handleKeydown,
  };
  return (
    <Hexagon hidden={hidden} {...attributes} {...handlers}>
      {children}
    </Hexagon>
  );
};
GameCell.displayName = 'Cell';
export default GameCell;
