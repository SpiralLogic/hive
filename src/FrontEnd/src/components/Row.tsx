import { deepEqual } from 'fast-equals';
import { FunctionComponent, h } from 'preact';
import { memo } from 'preact/compat';
import { HexCoordinates } from '../domain';
import Cell from './Cell';

type CellProps = typeof Cell.arguments['props'];
const cellKey = ({ q, r }: HexCoordinates) => `${q}-${r}`;
const createCell = (cell: CellProps) => <Cell key={cellKey(cell.coords)} {...cell} />;
const createPlaceholder = (key: number) => <div key={key} className="hidden" />;

type Props = {
    id: number;
    row: Array<CellProps>;
};

const Row: FunctionComponent<Props> = (props: Props) => {
    const { row } = props;
    const cells = row.map((cell, i) => (cell && createCell(cell)) || createPlaceholder(i));

    return <div className="hex-row">{cells}</div>;
};

Row.displayName = 'Row';
export default memo(Row, deepEqual);
