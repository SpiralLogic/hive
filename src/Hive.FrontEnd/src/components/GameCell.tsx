import { FunctionComponent } from 'preact';
import { useEffect, useState } from 'preact/hooks';
import { HexCoordinates, Tile as TileType } from '../domain';
import { TileEvent } from '../services';
import { addHiveDispatchListener, dispatchHiveEvent } from '../utilities/dispatcher';
import { handleDragOver, handleKeyboardNav, isEnterOrSpace } from '../utilities/handlers';
import { useClassReducer } from '../utilities/class-reducer';
import Hexagon from './Hexagon';
const isValidMove = (validMoves: HexCoordinates[], coords: HexCoordinates) =>
  validMoves.some((dest) => dest.q == coords.q && dest.r == coords.r);

type Props = { coords: HexCoordinates; hidden?: boolean };
const GameCell: FunctionComponent<Props> = (props) => {
  const { coords, children, hidden } = props;
  const [classes, setClasses] = useClassReducer('hide');
  const [selectedTile, setSelectedTile] = useState<TileType | null>(null);

  useEffect(() => setClasses({ type: hidden ? 'add' : 'remove', classes: ['hide'] }), [hidden]);

  addHiveDispatchListener<TileEvent>('tileDeselected', () => {
    setSelectedTile(null);
    setClasses({ type: 'remove', classes: ['active', 'can-drop', 'no-drop'] });
  });

  addHiveDispatchListener<TileEvent>('tileSelected', (intent) => {
    if (!isValidMove(intent.tile.moves, coords)) {
      setClasses({ type: 'add', classes: ['no-drop'] });
    } else {
      setSelectedTile(intent.tile);
      setClasses({ type: 'add', classes: ['can-drop'] });
    }
  });

  addHiveDispatchListener<TileEvent>('tileDropped', () => {
    if (classes.includes('active') && selectedTile && isValidMove(selectedTile.moves, coords)) {
      dispatchHiveEvent({
        type: 'move',
        move: { coords, tileId: selectedTile.id },
      });
    }
  });

  const handleDragLeave = (event: DragEvent) => {
    event.stopPropagation();
    setClasses({ type: 'remove', classes: ['active'] });
  };

  const handleDragEnter = () => {
    setClasses({ type: 'add', classes: ['active'] });
  };

  const handleClick = (event: UIEvent) => {
    if (!(selectedTile && isValidMove(selectedTile.moves, coords))) return;
    event.stopPropagation();
    dispatchHiveEvent({ type: 'tileClear', tile: selectedTile });
    dispatchHiveEvent({ type: 'move', move: { coords, tileId: selectedTile.id } });
  };

  const handleKeydown = (e: KeyboardEvent) => {
    if (handleKeyboardNav(e) || !isEnterOrSpace(e)) return;
    handleClick(e);
  };

  const attributes = {
    class: classes || undefined,
    tabindex: selectedTile && isValidMove(selectedTile.moves, coords) ? 0 : undefined,
    role: hidden ? 'none' : 'cell',
  };

  const handlers = {
    ondragover: handleDragOver,
    ondragleave: handleDragLeave,
    ondragenter: handleDragEnter,
    onmouseenter: handleDragEnter,
    onmouseleave: handleDragLeave,
    onclick: handleClick,
    onkeydown: handleKeydown,
  };
  return (
    <Hexagon hidden={hidden} {...attributes} {...handlers}>
      {children}
    </Hexagon>
  );
};
GameCell.displayName = 'Cell';
export default GameCell;
