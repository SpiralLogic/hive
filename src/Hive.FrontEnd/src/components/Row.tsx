import '../css/row.css';

import { FunctionComponent } from 'preact';

type Props = { class?: string; zIndex?: number; hidden?: boolean };

const Row: FunctionComponent<Props> = (props) => {
  const { hidden, zIndex } = props;
  return (
    <div class={props.class} role={hidden ? 'none' : 'row'} style={zIndex ? { zIndex } : undefined}>
      {props.children}
    </div>
  );
};

Row.displayName = 'Row';
export default Row;
