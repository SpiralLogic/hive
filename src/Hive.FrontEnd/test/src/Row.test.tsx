import { render, screen } from '@testing-library/preact';
import GameCell from '../../src/components/GameCell';
import Row from '../../src/components/Row';

describe('<Row>', () => {
  const row = () =>
    render(
      <Row>
        <GameCell coords={{ q: 0, r: 1 }} hidden={true} />
        <GameCell coords={{ q: 1, r: 1 }} />
        <GameCell coords={{ q: 2, r: 1 }} />
        <GameCell coords={{ q: 0, r: 1 }} hidden={true} />
      </Row>
    );

  it('renders multiple cells', () => {
    row();
    expect(screen.getAllByRole('cell')).toHaveLength(2);
    expect(screen.getAllByRole('none')).toHaveLength(2);
  });

  it('renders', () => {
    row();
    expect(screen.getByRole('row')).toMatchSnapshot();
  });
});
