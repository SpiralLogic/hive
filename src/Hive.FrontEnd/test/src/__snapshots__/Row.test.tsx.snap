// Vitest Snapshot v1

exports[`<Row> > renders 1`] = `
<div
  role="row"
>
  <div
    class="hide"
    role="none"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#hex"
      />
    </svg>
  </div>
  <div
    class=""
    role="cell"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#hex"
      />
    </svg>
  </div>
  <div
    class=""
    role="cell"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#hex"
      />
    </svg>
  </div>
  <div
    class="hide"
    role="none"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#hex"
      />
    </svg>
  </div>
</div>
`;

exports[`<Row> renders 1`] = `
<div
  role="row"
>
  <div
    class="hide"
    role="none"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#hex"
      />
    </svg>
  </div>
  <div
    class=""
    role="cell"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#hex"
      />
    </svg>
  </div>
  <div
    class=""
    role="cell"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#hex"
      />
    </svg>
  </div>
  <div
    class="hide"
    role="none"
  >
    <svg
      viewBox="0 0 100 100"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#hex"
      />
    </svg>
  </div>
</div>
`;
